### PR TITLE
Pr 11547

### LINKED ISSUE #1934: Ability to react to progress events of $http XHR
I would like to be able to bind to the progress event of the xhr upload.

for example (in plain javascript):

var xhr = new XMLHttpRequest();
xhr.upload.addEventListener("progress", uploadProgress, false)

(so I can show a progress meter for uploaded files)

I want to be able to do the same thing by doing $http.post...

for this and any other case edge case where you need to interact with the raw XHR object itself it might be nice to provide something that lets you just get at the original xhr object.

### PR DESCRIPTION
**Pull request by petebacondarwin**
**What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)**

**What is the current behavior? (You can also link to an open issue here)**

Closes #11547
Closes #1934

**What is the new behavior (if this is a feature change)?**

**Does this PR introduce a breaking change?**

**Please check if the PR fulfills these requirements**
- [ ] The commit message follows our guidelines: https://github.com/angular/angular.js/blob/master/CONTRIBUTING.md#commit-message-format
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**Other information**:

### CODE REVIEW

File: src/ngMock/angular-mocks.js [hunk 1]
```diff
@@ -2006,6 +2009,17 @@ function MockXhr() {
   };
 
   this.abort = angular.noop;
+
+  this.$$events = {};
+  this.addEventListener = function(name, listener) {
+    if (angular.isUndefined(this.$$events[name])) this.$$events[name] = [];
+    this.$$events[name].push(listener);
+  };
```
Comment (gkalpak) [inline src/ngMock/angular-mocks.js]: I am kind of confused with this snippet:

When we create a `MockXhr`, we set `this.$$events = {}`.
Then inside the mock `$httpBackend`, right after we create `xhr = new MockXhr()`, we overwrite `xhr.$$events = eventHandlers` (where `eventHandlers` is of type `{[key: string]: function}`).
Then, when `addEventListener` is called on a `MockXhr` instance, we check create an array on `this.$$events[name]` and push the listener.

I would expect the tests to fail, but they obviously don't, so what do I miss ?

Comment (petebacondarwin) [inline src/ngMock/angular-mocks.js]: I agree this is not optimal.
The `MockXhr` is used in two different scenarios:
* In the `$httpBackend` tests we instantiate `MockXhr` objects and pass them through to the `$httpBackend` service, which does call `addEventListener`. This allows us to test the "real" use of the `$httpBackend` service.
* In the `$http` tests, we have a mock `$httpBackend` service (from `ngMock`), which happens to hold its own instance of `MockXhr`. In these tests we never call `addEventListener` and I just chose to use the same `$$events` object for holding our mock event handlers for the test

Comment (gkalpak) [inline src/ngMock/angular-mocks.js]: I see :confused:
<sub>(Clarification comments welcome :smile:</sub>

**Comment by gkalpak** [comment]
LGTM

**Comment by benjamingr** [comment]
I'd like to see more tests for this but overall nice work :)

**Comment by petebacondarwin** [comment]
@benjamingr - what further tests would you like to see?

**Comment by benjamingr** [comment]
Well, the only test right now is that events are passed in one way - there is a single test, I'd like to se more coverage.

I'm also not sure how users are supposed to mock it, that should probably be documented.

**Comment by petebacondarwin** [comment]
@benjamingr could you be more explicit. What code paths are not being covered?

**Comment by petebacondarwin** [comment]
@benjamingr Can you create a new issue for improving the documentation of mocking the events so that we can track it?

**Pull request merged**